src/gh_pr_upsert/git.py

```diff
@@ -0,0 +1,223 @@
+"""Helpers for working with Git and GitHub."""
+from dataclasses import dataclass, field
+from functools import cache
+from subprocess import CalledProcessError
+from typing import Optional
+
+from gh_pr_upsert.run import run
+
+
+@dataclass(frozen=True)
+class User:
+    name: str
+    email: str
+
+
+@dataclass(frozen=True)
+class Commit:
+    sha: str
+    author: User
+
+    @classmethod
+    @cache
+    def get(cls, sha: str):
+        def git_show(format_string: str) -> str:
+            return run(["git", "show", "--no-patch", f"--format={format_string}", sha])
+
+        # `git show` doesn't have a machine-readable output mode (like a --json
+        # mode) and I don't fancy writing code to parse its human-readable
+        # output. So call `git show` multiple times with a different --format
+        # string each time to extract the different attributes of the commit.
+        return cls(
+            sha=git_show("%H"),  # Call git to make sure we get the full SHA.
+            author=User(
+                name=git_show("%an"),
+                email=git_show("%ae"),
+            ),
+        )
```

Comment:
You might think you can do this all at once with something like `git_show("%H, %an, %ae")` but there's nothing to prevent the author name and email from containing a comma and breaking this. I think it's more robust to run `git show` several times